### Northwind Traders Analytics

- Connect to PostgreSQL database and load SQL extension
- Perform data exploration to view all base tables
- Create tables with common joins for quering throughout the rest of the project

In [4]:
%load_ext sql
%sql postgresql://postgres@localhost:5432/northwind

In [5]:
%%sql

SELECT table_name, table_type 
  FROM information_schema.tables 
 WHERE table_schema = 'public';

 * postgresql://postgres@localhost:5432/northwind
17 rows affected.


[('territories', 'BASE TABLE'),
 ('order_details', 'BASE TABLE'),
 ('employee_territories', 'BASE TABLE'),
 ('us_states', 'BASE TABLE'),
 ('customers', 'BASE TABLE'),
 ('orders', 'BASE TABLE'),
 ('employees', 'BASE TABLE'),
 ('shippers', 'BASE TABLE'),
 ('products', 'BASE TABLE'),
 ('categories', 'BASE TABLE'),
 ('suppliers', 'BASE TABLE'),
 ('region', 'BASE TABLE'),
 ('customer_demographics', 'BASE TABLE'),
 ('customer_customer_demo', 'BASE TABLE'),
 ('customer_orders', 'VIEW'),
 ('product_quantity_orders', 'VIEW'),
 ('employee_orders', 'VIEW')]

In [119]:
%%sql 

CREATE OR REPLACE VIEW customer_orders AS 
SELECT c.*, o.order_id, o.employee_id, o.order_date, o.required_date, o.shipped_date, o.ship_via, o.freight, o.ship_name, o.ship_address, o.ship_city, o.ship_region, o.ship_postal_code, o.ship_country
  FROM customers AS c
  JOIN orders AS o 
    ON c.customer_id = o.customer_id;

CREATE OR REPLACE VIEW product_quantity_orders AS
SELECT p.*, od.quantity, od.discount, o.*
  FROM products AS p
  JOIN order_details AS od ON p.product_id = od.product_id
  JOIN orders AS o ON od.order_id = o.order_id
 ORDER BY p.product_id;

CREATE OR REPLACE VIEW employee_orders AS 
SELECT e.last_name, e.first_name, e.title, e.title_of_courtesy, e.birth_date, e.hire_date, address, city, region, postal_code, country, home_phone, extension, notes, reports_to, photo_path, o.*
  FROM employees AS e
  JOIN orders AS o ON e.employee_id = o.employee_id
 ORDER BY o.employee_id;

 * postgresql://postgres@localhost:5432/northwind
Done.
Done.
Done.


[]

#### Employees Sales Performance

- Rank employees based on their total sales amount

In [21]:
%%sql

WITH aggregate_sales AS(
SELECT e.employee_id, SUM(od.unit_price * od.quantity * (1 - od.discount)) AS total_sales
  FROM employees AS e
  JOIN orders AS o ON e.employee_id = o.employee_id
  JOIN order_details AS od ON o.order_id = od.order_id
 GROUP BY e.employee_id)

SELECT employee_id, ROUND(total_sales::numeric,2) AS total_sales, RANK() OVER(ORDER BY total_sales DESC) AS employee_rank
  FROM aggregate_sales;

 * postgresql://postgres@localhost:5432/northwind
9 rows affected.


[(4, Decimal('232890.85'), 1),
 (3, Decimal('202812.84'), 2),
 (1, Decimal('192107.60'), 3),
 (2, Decimal('166537.76'), 4),
 (8, Decimal('126862.28'), 5),
 (7, Decimal('124568.23'), 6),
 (9, Decimal('77308.07'), 7),
 (6, Decimal('73913.13'), 8),
 (5, Decimal('68792.28'), 9)]

- Calculated the total sales for each employee by multiplying unit price * quantity and subtracting any discounts at the order level
- Utilized ranking window function to identify high and low performers
- High performers are employees 4,3,1 and low performers are employees 9,6,5

#### Monthly Sales Trends

- Visualize the progress of the sales and identify trends that might shape the company's future strategies
- Aggregate sales data at a monthly level and calculating a running total of sales by month
- Analyze the month-over-month sales growth rate

In [117]:
%%sql

WITH monthly_sales AS (
SELECT DATE_TRUNC('month', order_date) AS order_month, SUM(unit_price * quantity * (1 - discount)) AS monthly_sales
  FROM orders 
  JOIN order_Details ON orders.order_id = order_details.order_id
 GROUP BY order_month
 ORDER BY order_month ASC)

SELECT order_month, ROUND(monthly_sales::numeric,2) AS monthly_sales, 
       ROUND(LAG(monthly_sales) OVER w::numeric,2) AS previous_month_sales,
       ROUND((monthly_sales::numeric -  LAG(monthly_sales) OVER w)::numeric / LAG(monthly_sales) OVER w::numeric * 100,2) AS MoM_growth_rate,
       ROUND(SUM(monthly_sales) OVER w::numeric,2) AS running_total
  FROM monthly_sales
WINDOW w AS (ORDER BY order_month);

 * postgresql://postgres@localhost:5432/northwind
23 rows affected.


[(datetime.datetime(1996, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('27861.90'), None, None, Decimal('27861.90')),
 (datetime.datetime(1996, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('25485.28'), Decimal('27861.90'), Decimal('-8.53'), Decimal('53347.17')),
 (datetime.datetime(1996, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('26381.40'), Decimal('25485.28'), Decimal('3.52'), Decimal('79728.57')),
 (datetime.datetime(1996, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('37515.72'), Decimal('26381.40'), Decimal('42.21'), Decimal('117244.30')),
 (datetime.datetime(1996, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('45600.05'), Decimal('37515.72'), Decimal('21.55'), Decimal('162844.34')),
 (datetime.datetime(1996, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('45239.63'), Decimal('45600.05'), Decimal('-0.79'), Decimal('208083.97')),
 (datetime.datetime(1997, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('61258.07'), Decimal('45239.63'), Decimal('35.41'), Decimal('269342.04')),
 (datetime.datetime(1997, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('38483.63'), Decimal('61258.07'), Decimal('-37.18'), Decimal('307825.68')),
 (datetime.datetime(1997, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('38547.22'), Decimal('38483.63'), Decimal('0.17'), Decimal('346372.90')),
 (datetime.datetime(1997, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('53032.95'), Decimal('38547.22'), Decimal('37.58'), Decimal('399405.85')),
 (datetime.datetime(1997, 5, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('53781.29'), Decimal('53032.95'), Decimal('1.41'), Decimal('453187.14')),
 (datetime.datetime(1997, 6, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('36362.80'), Decimal('53781.29'), Decimal('-32.39'), Decimal('489549.94')),
 (datetime.datetime(1997, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('51020.86'), Decimal('36362.80'), Decimal('40.31'), Decimal('540570.80')),
 (datetime.datetime(1997, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('47287.67'), Decimal('51020.86'), Decimal('-7.32'), Decimal('587858.47')),
 (datetime.datetime(1997, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('55629.24'), Decimal('47287.67'), Decimal('17.64'), Decimal('643487.71')),
 (datetime.datetime(1997, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('66749.23'), Decimal('55629.24'), Decimal('19.99'), Decimal('710236.94')),
 (datetime.datetime(1997, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('43533.81'), Decimal('66749.23'), Decimal('-34.78'), Decimal('753770.74')),
 (datetime.datetime(1997, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('71398.43'), Decimal('43533.81'), Decimal('64.01'), Decimal('825169.17')),
 (datetime.datetime(1998, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('94222.11'), Decimal('71398.43'), Decimal('31.97'), Decimal('919391.28')),
 (datetime.datetime(1998, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('99415.29'), Decimal('94222.11'), Decimal('5.51'), Decimal('1018806.57')),
 (datetime.datetime(1998, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('104854.16'), Decimal('99415.29'), Decimal('5.47'), Decimal('112366

#### High Value Customers
- Identify customers with above-average order values
- Rank customers based on the quantity of above-average orders and highlight the top 5 customers
- Identify the top 20% of customers by total purchase volume.

In [63]:
%%sql

WITH customer_order_totals AS (
SELECT c.customer_id, c.order_id, SUM(o.unit_price * o.quantity * (1 - o.discount)) AS order_total
  FROM customer_orders AS c
  JOIN order_details AS o ON c.order_id = o.order_id
 GROUP BY c.customer_id, c.order_id
 ORDER BY c.customer_id, c.order_id),

order_categorization AS (
SELECT customer_id, order_id, order_total,
       CASE
        WHEN order_total > AVG(order_total) OVER w THEN 'Above Average'
        WHEN order_total <= AVG(order_total) OVER w THEN 'Average/Below Average'
       END AS order_category
  FROM customer_order_totals
WINDOW w AS (ORDER BY order_total RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)),

customer_ranking AS (
SELECT customer_id, COUNT(DISTINCT order_id) AS above_avg_order_count,
       ROW_NUMBER() OVER(ORDER BY COUNT(DISTINCT order_id) DESC) AS customer_rnk
  FROM order_categorization
 WHERE order_category = 'Above Average'
 GROUP BY customer_id)

SELECT customer_id, above_avg_order_count
  FROM customer_ranking
 WHERE customer_rnk <=5;

 * postgresql://postgres@localhost:5432/northwind
5 rows affected.


[('ERNSH', 26), ('SAVEA', 26), ('QUICK', 22), ('HUNGO', 11), ('RATTC', 10)]

The top 5 customers with the most above-average orders are ERNSH, SAVEA, QUICK, HUNGO & RATTC.

#### Product and Category Performance
- Assess the sales composition across different product categories by finding the percent of total sales for each

In [123]:
%%sql

WITH aggregate_sales AS (
SELECT p.category_id, c.category_name, ROUND(SUM(unit_price * quantity * (1 - discount))::numeric,2) AS total_sales
  FROM product_quantity_orders AS p
  JOIN categories AS c ON p.category_id = c.category_id
 GROUP BY p.category_id, c.category_name
 ORDER BY total_sales DESC)

SELECT category_id, category_name, total_sales, ROUND(total_sales / (SELECT SUM(total_sales) FROM aggregate_sales) * 100.0,2) AS perc_of_total,
       SUM(total_sales) OVER(ORDER BY total_sales DESC) AS running_total,
       SUM(ROUND(total_sales / (SELECT SUM(total_sales) FROM aggregate_sales) * 100.0,2)) OVER(ORDER BY total_sales DESC) AS running_percent
  FROM aggregate_sales
 ORDER BY total_sales DESC;

 * postgresql://postgres@localhost:5432/northwind
8 rows affected.


[(1, 'Beverages', Decimal('288694.65'), Decimal('21.33'), Decimal('288694.65'), Decimal('21.33')),
 (4, 'Dairy Products', Decimal('251147.60'), Decimal('18.56'), Decimal('539842.25'), Decimal('39.89')),
 (3, 'Confections', Decimal('179863.65'), Decimal('13.29'), Decimal('719705.90'), Decimal('53.18')),
 (6, 'Meat/Poultry', Decimal('174622.55'), Decimal('12.90'), Decimal('894328.45'), Decimal('66.08')),
 (8, 'Seafood', Decimal('137989.31'), Decimal('10.20'), Decimal('1032317.76'), Decimal('76.28')),
 (2, 'Condiments', Decimal('113692.19'), Decimal('8.40'), Decimal('1146009.95'), Decimal('84.68')),
 (7, 'Produce', Decimal('105745.70'), Decimal('7.81'), Decimal('1251755.65'), Decimal('92.49')),
 (5, 'Grains/Cereals', Decimal('101646.94'), Decimal('7.51'), Decimal('1353402.59'), Decimal('100.00'))]

The most popular product categories are beverages, dairy products and confections which make up 53.18% of all sales ($719,705.90)

#### Top Products Per Category
- Drill down further into each group to determine the top three items sold in each product category

In [93]:
%%sql

WITH sales_by_product AS (
SELECT c.category_name, p.product_name, ROUND(SUM(unit_price * quantity * (1 - discount))::numeric,2) AS total_sales,
       ROW_NUMBER() OVER(PARTITION BY c.category_name ORDER BY SUM(unit_price * quantity * (1 - discount))::numeric DESC) AS product_ranking
  FROM product_quantity_orders AS p
  JOIN categories AS c ON p.category_id = c.category_id
 GROUP BY 1,2)

SELECT category_name, product_name, total_sales
  FROM sales_by_product
 WHERE product_ranking <= 3

 * postgresql://postgres@localhost:5432/northwind
24 rows affected.


[('Beverages', 'Côte de Blaye', Decimal('153897.17')),
 ('Beverages', 'Ipoh Coffee', Decimal('25109.10')),
 ('Beverages', 'Chang', Decimal('17719.40')),
 ('Condiments', 'Vegie-spread', Decimal('18343.62')),
 ('Condiments', "Sirop d'érable", Decimal('15022.35')),
 ('Condiments', 'Louisiana Fiery Hot Pepper Sauce', Decimal('14893.93')),
 ('Confections', 'Tarte au sucre', Decimal('50737.09')),
 ('Confections', "Sir Rodney's Marmalade", Decimal('24199.56')),
 ('Confections', 'Gumbär Gummibärchen', Decimal('21662.69')),
 ('Dairy Products', 'Raclette Courdavault', Decimal('76683.75')),
 ('Dairy Products', 'Camembert Pierrot', Decimal('49877.32')),
 ('Dairy Products', 'Mozzarella di Giovanni', Decimal('27086.58')),
 ('Grains/Cereals', 'Gnocchi di nonna Alice', Decimal('45351.10')),
 ('Grains/Cereals', 'Wimmers gute Semmelknödel', Decimal('23487.47')),
 ('Grains/Cereals', 'Singaporean Hokkien Fried Mee', Decimal('8986.60')),
 ('Meat/Poultry', 'Thüringer Rostbratwurst', Decimal('84783.77')),
 ('Meat/Poultry', 'Alice Mutton', Decimal('35105.85')),
 ('Meat/Poultry', 'Perth Pasties', Decimal('22623.80')),
 ('Produce', 'Manjimup Dried Apples', Decimal('43846.90')),
 ('Produce', 'Rössle Sauerkraut', Decimal('27936.84')),
 ('Produce', "Uncle Bob's Organic Dried Pears", Decimal('22453.50')),
 ('Seafood', 'Carnarvon Tigers', Decimal('30728.12')),
 ('Seafood', 'Ikura', Decimal('21653.50')),
 ('Seafood', 'Boston Crab Meat', Decimal('19055.04'))]

#### Additional Metrics
- Compare each employee's sales performance to the average sales performance of all employees to evaluate their contributions.
- Analyze the cumulative sales for each product category over a specified period to understand each category's growth.
- Identify products that perform exceptionally well in terms of sales. Specifically, products whose total sales exceed the 75th percentile.